lib/TableGen/DialectType.cpp

```diff
@@ -195,76 +270,141 @@ void DialectType::emitDefinition(raw_ostream &out, GenDialect *dialect) const {
   fmt.addSubst("ints", symbols.chooseName("ints"));
   fmt.addSubst("_errs", symbols.chooseName("errs"));
 
-  // Output the type argument getters.
-  unsigned typeIdx = 0;
-  unsigned intIdx = 0;
-  for (const auto &argument : typeArguments()) {
-    std::string expr;
-    if (argument.type->isTypeArg()) {
-      expr = tgfmt("type_params()[$0]", &fmt, typeIdx);
-      ++typeIdx;
-    } else {
-      expr = tgfmt("int_params()[$0]", &fmt, intIdx);
-      expr = tgfmt(cast<Attr>(argument.type)->getFromUnsigned(), &fmt, expr);
-      ++intIdx;
+  if (m_structBacked) {
+    out << tgfmt("$_type* $_type::get(", &fmt);
+    bool contextPresent =
+        !m_getterArguments.empty() && m_getterArguments.front().cppType.find(
+                                          "LLVMContext") != std::string::npos;
+    if (!contextPresent) {
+      out << "::llvm::LLVMContext &" << m_context;
+      if (!m_getterArguments.empty())
+        out << ", ";
+    }
+    for (auto argument : llvm::enumerate(m_getterArguments)) {
+      if (argument.index() != 0)
+        out << ", ";
+      out << argument.value().cppType << ' ' << argument.value().name;
+    }
+    out << ") {\n";
+
+    auto getterArgs =
+        ArrayRef<GetterArg>(m_getterArguments).drop_front(m_argBegin);
+
+    for (const auto &[argument, getterArg] :
+         llvm::zip(typeArguments(), getterArgs)) {
+      if (auto *attr = dyn_cast<Attr>(argument.type)) {
+        out << tgfmt(attr->getCheck(), &fmt, getterArg.name) << '\n';
+      }
     }
 
-    FmtContextScope scope{fmt};
-    fmt.addSubst("type", argument.type->getCppType());
-    fmt.addSubst("name", convertToCamelFromSnakeCase(argument.name, true));
-    fmt.addSubst("expr", expr);
+    out << "  std::string __name; ::llvm::raw_string_ostream __os(__name);\n";
+    out << tgfmt("  __os << \"$0\";\n", &fmt, m_structPrefix);
+    for (const auto &getterArg : getterArgs)
+      out << "  __os << (uint64_t)" << getterArg.name << " << '.';\n";
```

Comment:
Slightly nicer IR output:
```suggestion
    for (const auto &getterArg : getterArgs)
      out << "  __os << '.' << (uint64_t)" << getterArg.name << ";\n";
```